falcon/testing/client.py

```diff
@@ -485,51 +700,319 @@ def simulate_request(app, method='GET', path='/', query_string=None,
                   code_to_http_status(resp_event_collector.status),
-                  resp_event_collector.headers,
-                  events=resp_event_collector.events)
+                  resp_event_collector.headers)
+
+
+class ASGIConductor:
+    """Test conductor for ASGI apps.
+
+    This class provides more control over the lifecycle of a simulated
+    request as compared to :class:`~.TestClient`. In addition, the conductor's
+    asynchronous interface affords interleaved requests and the testing of
+    streaming protocols such as server-sent events (SSE) and WebSocket.
+
+    :class:`~.ASGIConductor` is implemented as a context manager. Upon
+    entering and exiting the context, the appropriate ASGI lifespan events
+    will be simulated.
+
+    Within the context, HTTP requests can be simulated using an interface
+    that is similar to :class:`~.TestClient`, except that all ``simulate_*()``
+    methods are coroutines::
+
+        async with testing.ASGIConductor(some_app) as conductor:
+            async def post_events():
+                for i in range(100):
+                    await conductor.simulate_post('/events', json={'id': i}):
+                    await asyncio.sleep(0.01)
+
+            async def get_events_sse():
+                # Here, we will get only some of the single server-sent events
+                # because the non-streaming method is "single-shot". In other
+                # words, simulate_get() will emit a client disconnect event
+                # into the app before returning.
+                result = await conductor.simulate_get('/events')
+
+                # Alternatively, we can use simulate_get_stream() as a context
+                # manager to perform a series of reads on the result body that
+                # are interleaved with the execution of the post_events()
+                # coroutine.
+                async with conductor.simulate_get_stream('/events') as sr:
+                    while some_condition:
+                        # Read next body chunk that was received (if any).
+                        chunk = await sr.stream.read()
+
+                        if chunk:
+                            # TODO: Do something with the chunk
+                            pass
+
+                    # Exiting the context causes the request event emitter to
+                    # begin emitting 'http.disconnect' events and then awaits
+                    # the completion of the asyncio task that is running the
+                    # simulated ASGI request.
+
+            asyncio.gather(post_events(), get_events_sse())
+
+    Note:
+        Because the :class:`~.ASGIConductor` interface uses coroutines,
+        it cannot be used directly with synchronous testing frameworks such as
+        pytest. As a workaround, the test can be adapted by wrapping it in
+        an inline async function and then invoking it via
+        :meth:`falcon.invoke_coroutine_sync`. Alternatively, you can try
```

Comment:
To that end, I've also added the [`runs_sync`](https://falcon.readthedocs.io/en/latest/api/util.html#falcon.runs_sync) decorator.